### DISCUSSION TITLE
When to Unsubscribe?

### DISCUSSION THREAD

**Discussion by R2rho** [pmndrs/valtio#583] [Q&A]
Similar to #449 

I've been utilizing subscribe outside of React to maintain state in separate state objects needed from a query like this

```javascript


export const state = proxy({Obj1:{key1:someData, key2:someOtherData},
                            Obj2:{key1:someData, key2:{nestedKey1:someOtherData}}, 
                            })

const unsubscribe = subscribe(state.Obj1, async ()=>{
    //Fetch new data for Obj2 based on state changes in Obj1
    const newData = await fetchNewObj2Data(state.obj1)
    //Update obj2
    for (const key in state.Obj2){ delete state.obj2[key]}
    Object.assign(state.Obj2, await newData)

})
```

Two questions:
1) I'm not sure I understand the use of unsubscribe outside the context of a useEffect return callback. Since this is outside of react, what happens to the subscription if state.Obj1 were to get deleted? Do I need to call unsubscribe() somewhere, or does the subsciption get garbage collected somehow?

2) If I'm using useSnapshot on Obj2 in a React component to display the updated data, what happens to components that I took a snapshot on state.Obj2.key2? If I'm passing down as props, I'm assuming the component remounts and a new snapshot is taken on the new data?

Just trying to understand the behavior so I don't make assumptions as to how the application is behaving.

**Comment by dai-shi** [comment]
1: It will be garbage collected.

2: snapshots are immutable objects. They don't trigger re-renders. `useSnapshot` will. Generally, you want to prefer passing states (not snaps) as props.
